File: leetcode/3559-number-of-ways-to-assign-edge-weights-ii/3559-number-of-ways-to-assign-edge-weights-ii.kt
class Solution {

    companion object {
        private const val MOD = 1_000_000_007L
        private const val LOG = 18
    }

    fun assignEdgeWeights(
        edges: Array<IntArray>,
        queries: Array<IntArray>
    ): IntArray {

        val n = edges.size + 1

        // Used adjacency list for the tree.
        val graph = Array(n + 1) {
            mutableListOf<Int>()
        }

        for (edge in edges) {

            val u = edge[0]
            val v = edge[1]

            graph[u].add(v)
            graph[v].add(u)
        }

        // Used depth to store node depths.
        val depth = IntArray(n + 1)

        // Used binary lifting table.
        val up = Array(LOG) {
            IntArray(n + 1)
        }

        // Built depths and parents with BFS.
        val queue = ArrayDeque<Int>()

        val visited = BooleanArray(n + 1)

        queue.addLast(1)
        visited[1] = true

        while (queue.isNotEmpty()) {

            val node = queue.removeFirst()

            for (next in graph[node]) {

                if (visited[next]) {
                    continue
                }

                visited[next] = true

                depth[next] = depth[node] + 1

                up[0][next] = node

                queue.addLast(next)
            }
        }

        // Built binary lifting table.
        for (k in 1 until LOG) {

            for (node in 1..n) {

                up[k][node] =
                    up[k - 1][
                        up[k - 1][node]
                    ]
            }
        }

        // Precomputed powers of 2 modulo MOD.
        val pow2 = IntArray(n + 1)

        pow2[0] = 1

        for (i in 1..n) {

            pow2[i] =
                ((pow2[i - 1].toLong() * 2) % MOD)
                    .toInt()
        }

        // Used LCA to find lowest common ancestor.
        fun lca(
            startA: Int,
            startB: Int
        ): Int {

            var a = startA
            var b = startB

            if (depth[a] < depth[b]) {
                val temp = a
                a = b
                b = temp
            }

            val diff = depth[a] - depth[b]

            for (k in 0 until LOG) {

                if ((diff and (1 shl k)) != 0) {

                    a = up[k][a]
                }
            }

            if (a == b) {
                return a
            }

            for (k in LOG - 1 downTo 0) {

                if (up[k][a] != up[k][b]) {

                    a = up[k][a]
                    b = up[k][b]
                }
            }

            return up[0][a]
        }

        // Processed all queries.
        val answer = IntArray(queries.size)

        for (i in queries.indices) {

            val u = queries[i][0]
            val v = queries[i][1]

            val ancestor = lca(u, v)

            val dist =
                depth[u] +
                depth[v] -
                2 * depth[ancestor]

            answer[i] =
                if (dist == 0) {
                    0
                } else {
                    pow2[dist - 1]
                }
        }

        return answer
    }
}

File: leetcode/3559-number-of-ways-to-assign-edge-weights-ii/3559-number-of-ways-to-assign-edge-weights-ii.cpp
class Solution {
public:
    static constexpr long long MOD = 1'000'000'007;

    vector<int> assignEdgeWeights(
        vector<vector<int>>& edges,
        vector<vector<int>>& queries
    ) {
        int n = edges.size() + 1;

        vector<vector<int>> graph(n + 1);

        for (auto& e : edges) {
            int u = e[0];
            int v = e[1];

            graph[u].push_back(v);
            graph[v].push_back(u);
        }

        int LOG = 18;

        vector<int> depth(n + 1, 0);

        vector<vector<int>> up(
            LOG,
            vector<int>(n + 1)
        );

        queue<int> q;
        q.push(1);

        vector<bool> visited(n + 1, false);
        visited[1] = true;

        while (!q.empty()) {

            int node = q.front();
            q.pop();

            for (int next : graph[node]) {

                if (visited[next]) {
                    continue;
                }

                visited[next] = true;

                depth[next] = depth[node] + 1;

                up[0][next] = node;

                q.push(next);
            }
        }

        for (int k = 1; k < LOG; k++) {

            for (int node = 1; node <= n; node++) {

                up[k][node] =
                    up[k - 1][
                        up[k - 1][node]
                    ];
            }
        }

        vector<int> pow2(n + 1, 1);

        for (int i = 1; i <= n; i++) {

            pow2[i] =
                (2LL * pow2[i - 1]) % MOD;
        }

        auto lca = [&](int a, int b) {

            if (depth[a] < depth[b]) {
                swap(a, b);
            }

            int diff = depth[a] - depth[b];

            for (int k = 0; k < LOG; k++) {

                if (diff & (1 << k)) {
                    a = up[k][a];
                }
            }

            if (a == b) {
                return a;
            }

            for (int k = LOG - 1; k >= 0; k--) {

                if (up[k][a] != up[k][b]) {

                    a = up[k][a];
                    b = up[k][b];
                }
            }

            return up[0][a];
        };

        vector<int> answer;
        answer.reserve(queries.size());

        for (auto& query : queries) {

            int u = query[0];
            int v = query[1];

            int ancestor = lca(u, v);

            int dist =
                depth[u]
                + depth[v]
                - 2 * depth[ancestor];

            if (dist == 0) {
                answer.push_back(0);
            } else {
                answer.push_back(
                    pow2[dist - 1]
                );
            }
        }

        return answer;
    }
};